Register custom functions

josefalanga opened the pull request:
This PR implements partially issue #5

Will remain in draft until we define if we should implement blocking commands and functions as well (replacement for Unity coroutines)

josefalanga committed:
failing tests

josefalanga committed:
first approach

josefalanga committed:
refactor with nested functions for legibility

weaversam8 commented:
See my comment on https://github.com/relaypro-open/YarnRunner-Python/issues/5#issuecomment-1594880046.

I think this is great, a couple small things:

1. Do you mind adding a note to the README (probably in the example code somewhere) explaining the difference between `runner.add_command_handler(...)` and `runner.add_function_handler(...)`? I worry people might get confused.

2. Do you mind adding a test case for an unimplemented function, to ensure the error handling works as expected?

Thanks!

josefalanga committed:
documentation for commands and functions

josefalanga committed:
add test for exception handling with calls to unregistered custom functions

josefalanga commented:
I added documentation to both methods as Docstrings, with links to YS docs to minimize confussion about the two terms. Also I added to README some clarification about possible uses (adding missing built-in functions and maybe the `<<wait>>` command if neccesary)

josefalanga commented on yarnrunner_python/runner.py:
```diff
@@ -8,6 +10,10 @@
+        # reset files' position before reading
+        compiled_yarn_f.seek(0, 0)
+        names_csv_f.seek(0, 0)
+
```
This might be a little controversial, but it has some uses besides testing purposes. If you think it's bad, I can roll back this change and use this trick from outside the runner.

josefalanga replied:
Also I removed it from the testing cases of experimental lines.

weaversam8 commented on yarnrunner_python/runner.py:
```diff
@@ -147,8 +154,27 @@ def choose(self, index):
+        """
+        Registers a custom command that can be invoked from Yarn scripts.
+        Like `<<get_player_name>>` or `<<walk MyCharacter StageLeft>>`.
+        Usually these commands don't return anything. Returning a string, will put it on the line where is used.
```
> Returning a string, will put it on the line where is used.

Is this how Yarn handles this in Unity? I was under the impression that the return value had no bearing on the story and that this was a method to trigger events within the external "game".

josefalanga replied:
According to the documentation, you are correct. If you want to retrieve some value from outside yarn, you should use a Function, not a Command, all samples for command handlers are void methods in C#. I'm not sure what it does under the hood because you can register coroutines as commands as well, but in general terms, yeah, there shouldn't be a return value used anywhere.

I wrote that because that's how the runner is behaving right now without this PR. It checks if the value returned from a command is a string, to put it in the vm stack, and that gets printed to a line. You even tested that use case in the file `test_basic.py:test_start_node_choose2`. Now that we have the formal method of getting something from outside the runner via functions, maybe we should delete the old behaviour?

weaversam8 replied:
Ah, that's fine, I didn't know that was the current runner behavior. Thanks for catching that!

weaversam8 commented on tests/test_experimental_newlines.py:
```diff
@@ -7,13 +7,7 @@
-names_csv_f.seek(0, 0)
-
-runner_experimental = YarnRunner(
-    compiled_yarn_f, names_csv_f, experimental_newlines=True)
```
Was seeking to reset this file no longer necessary?

josefalanga replied:
That's the controversial part I mentioned in the other comment, in line 13 of `runner.py`. I moved this file reset to the constructor, so everytime you build a runner, it does with a valid state. I think there are other use cases for this besided testing, but maybe we should discuss it in another issue/PR, because this one was only for functions, and get this part reverted.

weaversam8 replied:
Ah, I didn't catch that you had moved this inside the runner. I think that's fine.

weaversam8 commented:
Your comments make sense, I'm ready to merge. Thanks for your hard work on this!
